packages/otel/src/otel-wrapper.ts

```diff
@@ -0,0 +1,124 @@
+
+const tracer = getOtelTracer();
+
+if (process.env.OTEL_ENABLED === "true" && process.env.OTEL_SERVICE_NAME) {
+  otelSdk.start();
+}
+
+const OTEL_FLUSH_TIMEOUT = sharedOtelConfig.flushTimeout;
+
+const flushOtel = async () => {
+  await race({
+    promise: loggerProvider.forceFlush(),
+    error: new Error("Timeout flushing OTEL logs from provider"),
+    timeout: OTEL_FLUSH_TIMEOUT,
+  });
+
+  await race({
+    promise: Promise.all([batchSpanProcessor.forceFlush(), otelLogsProcessor.forceFlush()]),
+    error: new Error("Timeout flushing OTEL items from processors"),
+    timeout: OTEL_FLUSH_TIMEOUT,
+  });
+};
+
+export const withOtel = (handler: NextApiHandler, staticRouteName: string): NextApiHandler => {
+  if (process.env.OTEL_ENABLED !== "true") {
+    return handler;
+  }
+
+  return new Proxy(handler, {
+    apply: async (
+      wrappingTarget,
+      thisArg,
+      args: [NextApiRequest | undefined, NextApiResponse | undefined],
+    ) => {
+      const [req, res] = args;
+
+      if (!req || !res) {
+        // winstonLogger.warn("No request and/or response objects found, OTEL is not set-up");
+
+        // @ts-expect-error runtime check
+        return wrappingTarget.apply(thisArg, args);
+      }
+
+      const attributesFromRequest = getAttributesFromRequest(req);
+
+      return tracer.startActiveSpan(
+        `${attributesFromRequest[SemanticAttributes.HTTP_METHOD]} ${staticRouteName}`,
+        {
+          kind: SpanKind.SERVER,
+          attributes: attributesFromRequest,
+        },
+        async (span) => {
+          span.setAttribute(SemanticAttributes.HTTP_ROUTE, staticRouteName);
+
+          const originalResEnd = res.end;
+
+          /**
+           * Override native res.end to flush OTEL traces before it ends
+           */
+          // @ts-expect-error - this is a hack to get around Vercel freezing lambda's
+          res.end = async function (this: unknown, ...args: unknown[]) {
+            span.setAttribute(SemanticAttributes.HTTP_STATUS_CODE, res.statusCode);
+            span.end();
+
+            try {
+              await flushOtel();
+            } catch (e) {
+              // winstonLogger.error("Failed to flush OTEL", { error: e });
+              // noop - don't block return even if we loose traces
+            }
+
+            // @ts-expect-error passthrough args to the original function
+            return originalResEnd.apply(this, args);
+          };
+
+          try {
+            // return await loggerContext.wrap(() => wrappingTarget.apply(thisArg, [req, res]));
+
+            wrappingTarget.apply(thisArg, [req, res]);
+          } catch (error) {
+            span.setAttribute(SemanticAttributes.HTTP_STATUS_CODE, 500);
+
+            setErrorOnSpan(error, span);
+
+            span.end();
+
+            try {
+              await flushOtel();
+            } catch (e) {
+              // winstonLogger.error("Failed to flush OTEL", { error: e });
```

Comment:
Can we uncomment this?

Reply:
I did, but fallback to console, because I don't have extracted logger yet. Will improve in the future